From 4f141b72ddf2b8c12b3914a83885faa711371618 Mon Sep 17 00:00:00 2001
From: Dmitry Petrochenko <dmitry.petrochenko@intel.com>
Date: Thu, 28 Mar 2013 13:44:30 +0700
Subject: Dalvik: Make conditions safe in VRDelay functions (Klockwork issue)

BZ: 96357

The "<= VRDELAY_COUNT" condition was not safe since it allowed
index "3" for array if size=3. Fixed to "< VRDELAY_COUNT"

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I8e3db5813675455c6b2ccfaaf15195d9e62ac7d3
Orig-MCG-Change-Id: If2d996838f6d9958cd5aaec140b96aa3d82acbbe
Signed-off-by: Dmitry Petrochenko <dmitry.petrochenko@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/AnalysisO1.cpp |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/vm/compiler/codegen/x86/AnalysisO1.cpp b/vm/compiler/codegen/x86/AnalysisO1.cpp
index 7d5705c..625a59e 100644
--- a/vm/compiler/codegen/x86/AnalysisO1.cpp
+++ b/vm/compiler/codegen/x86/AnalysisO1.cpp
@@ -4268,7 +4268,7 @@ int requestVRFreeDelay(int regNum, u4 reason) {
     // TODO Add 64-bit operand support when needed
     int indexL = searchMemTable(regNum);
     if(indexL >= 0) {
-        if(reason >= 0 && reason <= VRDELAY_COUNT) {
+        if(reason >= 0 && reason < VRDELAY_COUNT) {
 #ifdef DEBUG_REGALLOC
             ALOGI("requestFreeDelay: reason=%i VR=%d count=%i", reason, regNum, memVRTable[indexL].delayFreeCounters[reason]);
 #endif
@@ -4304,7 +4304,7 @@ int cancelVRFreeDelayRequest(int regNum, u4 reason) {
     bool needCallToFreeReg = false;
     int indexL = searchMemTable(regNum);
     if(indexL >= 0) {
-        if(reason >= 0 && reason <= VRDELAY_COUNT) { // don't cancel delay if it wasn't requested
+        if(reason >= 0 && reason < VRDELAY_COUNT) { // don't cancel delay if it wasn't requested
             if(memVRTable[indexL].delayFreeCounters[reason] > 0) {
 #ifdef DEBUG_REGALLOC
                 ALOGI("cancelVRFreeDelay: reason=%i VR=%d count=%i", reason, regNum, memVRTable[indexL].delayFreeCounters[reason]);
-- 
1.7.4.1

